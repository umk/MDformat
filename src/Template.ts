import { Placeholder, Ref, TemplateParts } from './TemplatePart'

const ID_REGEX = '[_\\p{L}][_\\p{L}\\p{Nd}]*'

/**
 * A regular expression that matches a sequence of JavaScript
 * property names separated by dots. This supports only a subset of
 * the JavaScript property names.
 */
const PATH_REGEX = `${ID_REGEX}(\\.${ID_REGEX})*`

/**
 * A regular expression that matches template data field reference. The
 * path denotes the path through object hierarchy to the data field, that
 * substitutes the reference. The group is optional and denotes a key,
 * that is used to zip the data field values with the same group value.
 */
const REF_REGEX = new RegExp(`\\{\\{((?<group>${ID_REGEX}):)?(?<path>${PATH_REGEX})\\}\\}`, 'gu')

type Template = {
  /**
   * The template parts.
   */
  parts: TemplateParts
  /**
   * A collection of fields referenced from the template
   * @internal
   */
  refs?: Array<Ref>
} & { __brand: 'template' }

export function getTemplateRefs(template: Template): Array<Ref> {
  let refs = template.refs
  if (!refs) {
    template.refs = refs = template.parts.filter((p) => typeof p === 'object') as Array<Ref>
  }
  return refs
}

export function createTemplate(template: string): Template {
  const matches = Array.from(template.matchAll(REF_REGEX)).map((match) => ({
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    index: match.index!,
    raw: match[0],
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    path: match.groups!.path,
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    group: match.groups!.group ?? match.groups!.path,
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    modifier: match.groups!.modifier as string | undefined,
  }))
  const parts: TemplateParts = []
  // const refs: Map<string, Ref> = new Map()
  const refs: Record<string, Record<string, Ref>> = {}
  let lastIndex = 0
  for (const { index, raw, path, group } of matches) {
    parts.push(template.slice(lastIndex, index))
    lastIndex = index + raw.length
    const properties = path.split('.')
    const ref: Placeholder = { path, properties, group, raw }
    ;(refs[group] || (refs[group] = {}))[path] = ref
    parts.push(ref)
  }
  parts.push(template.slice(lastIndex))
  return {
    parts,
    refs: Object.values(refs).flatMap((g) => Object.values(g)),
  } as Template
}

export default Template
