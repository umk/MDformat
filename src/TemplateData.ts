import { configuration } from './Configuration'
import Content, { ContentPrimitive } from './Content'
import Template, { getTemplateRefs } from './Template'
import { Ref } from './TemplatePart'

/**
 * A mapping from a group name to an object that contains
 * another mapping from a data reference field to substitution
 * for that field.
 */
type TemplateData = {
  groups: Record<string, TemplateDataGroup>
}

export type TemplateDataGroup = {
  /**
   * Index of the record in the group taken as a substitution
   * for placeholder.
   */
  index: number
  refs: TemplateDataRefs
}

/**
 * A mapping from a data reference field to an object that
 * contains substitution for that field. The key is provided
 * in exact wording that provided in the template.
 */
export type TemplateDataRefs = Record<string, Array<Content>>

export function getTemplateData(
  templateOrRefs: Template | Array<Ref>,
  content: Content,
): Array<TemplateData> {
  const refs = Array.isArray(templateOrRefs) ? templateOrRefs : getTemplateRefs(templateOrRefs)
  const refsByProperty = refs.reduce(
    (refByProperty, { path, properties: [property, ...rest], group }) => {
      let refGroup = refByProperty.get(property)
      if (!refGroup) {
        refByProperty.set(property, (refGroup = []))
      }
      refGroup.push({
        path,
        properties: rest,
        group,
      })
      return refByProperty
    },
    new Map<string | undefined, Array<Ref>>(),
  )
  const refsArr = Array.from(refsByProperty.entries())
  const items = resolveContent(content)
  return items.flatMap((item, index, values) =>
    Array.from(refsArr)
      .map(([property, samePropRefs]): Array<TemplateData> => {
        // If there is a property to traverse in order to reach
        // the placeholder value...
        if (property) {
          // ...if property value is an object...
          if (typeof item === 'object' && item !== null && property in item) {
            // ...get the substitution value from that property
            const data: Content = item[property]
            return getTemplateData(samePropRefs, data)
          } else {
            // ...otherwise substitution could not be found
            return [{ groups: {} }]
          }
        } else {
          // ...otherwise use current value as a placeholder substitution
          const onePropResult = samePropRefs.reduce<TemplateData>(
            (previous, { group, path }) => {
              let dataGroup = previous.groups[group]
              if (!dataGroup) {
                previous.groups[group] = dataGroup = { index, refs: {} }
              }
              dataGroup.refs[path] = values
              return previous
            },
            {
              groups: {},
            },
          )
          return [onePropResult]
        }
      })
      .sort((a, b) => {
        // Sort the results by the number of items for greater
        // possibility to shrink the number of results based on
        // groupings or missing properties.
        return a.length - b.length
      })
      .reduce(permutateData, [{ groups: {} }]),
  )
}

function permutateData(a: Array<TemplateData>, b: Array<TemplateData>): Array<TemplateData> {
  const result: Array<TemplateData> = []
  for (const itemA of a) {
    itemsB: for (const itemB of b) {
      const item: TemplateData = { groups: {} }
      const entries = [...Object.entries(itemA.groups), ...Object.entries(itemB.groups)]
      for (const [group, { index, refs }] of entries) {
        const groupData = item.groups[group]
        if (!groupData) {
          item.groups[group] = { index, refs: { ...refs } }
        } else {
          if (groupData.index !== index) {
            // Being in the same group, two properties have
            // different indexes and cannot be zipped.
            continue itemsB
          }
          item.groups[group] = { index, refs: { ...groupData.refs, ...refs } }
        }
      }
      result.push(item)
      if (result.length > configuration.maxPermutations) {
        throw new Error('The template resolves into too many combinations')
      }
    }
  }
  return result
}

function resolveContent(content: Content): Array<ContentPrimitive | Record<string, Content>> {
  if (Array.isArray(content)) {
    return content.flatMap((c) => resolveContent(c))
  }
  return [content]
}

export default TemplateData
