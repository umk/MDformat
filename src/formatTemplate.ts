import Template from './Template'
import TemplateData from './TemplateData'
import { isPlaceholder } from './TemplatePart'

/**
 * Creates a content that represents the template bound to the data
 * object, where each of the data object properties can be referenced
 * from the template.
 * @param template The template structure.
 * @param data A template data record
 * @returns A string that represents the template bound to the data object.
 */
function formatTemplate(template: Template, data: TemplateData): string {
  return template.parts
    .map((part) => {
      if (!isPlaceholder(part)) {
        return part
      }
      const { path, group } = part
      const d = data.groups[group]
      if (d) {
        const { index, refs } = data.groups[group]
        const ref = refs[path]
        if (ref) {
          let content = ref[index]
          if (part.formatter) {
            content = part.formatter(content)
          }
          return content
        }
      }
      return part.raw
    })
    .join('')
}

export default formatTemplate
