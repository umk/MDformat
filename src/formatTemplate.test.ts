import { faker } from '@faker-js/faker'

import { createTemplate } from './Template'
import { getTemplateData } from './TemplateData'
import formatTemplate from './formatTemplate'

describe('template', () => {
  it('creates template without data', () => {
    const template = createTemplate(faker.lorem.text())
    const data = getTemplateData(template, {})
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('creates template with single data field', () => {
    const template = createTemplate('Hello {{name}}!')
    const data = getTemplateData(template, {
      name: faker.person.firstName(),
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('creates template with single array data field', () => {
    const template = createTemplate('Hello {{name}}!')
    const data = getTemplateData(template, {
      name: [faker.person.firstName(), faker.person.firstName(), faker.person.firstName()],
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('creates template with two permutated array data fields', () => {
    const template = createTemplate('{{name}} goes to {{city}}!')
    const data = getTemplateData(template, {
      name: [faker.person.firstName(), faker.person.firstName(), faker.person.firstName()],
      city: [faker.location.city(), faker.location.city()],
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it("doesn't permutate properties, that are not referenced in template", () => {
    const template = createTemplate('Hello {{name}}!')
    const data = getTemplateData(template, {
      name: [faker.person.firstName(), faker.person.firstName(), faker.person.firstName()],
      city: [faker.location.city(), faker.location.city()],
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('creates template with two zipped array data fields', () => {
    const template = createTemplate('{{itinerary:name}} goes to {{itinerary:city}}!')
    const data = getTemplateData(template, {
      name: [faker.person.firstName(), faker.person.firstName(), faker.person.firstName()],
      city: [faker.location.city(), faker.location.city(), faker.location.city()],
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('creates template with three combined array data fields', () => {
    const template = createTemplate(
      '{{name}} goes to {{destination:city}} on {{destination:date}}!',
    )
    const data = getTemplateData(template, {
      name: [faker.person.firstName(), faker.person.firstName()],
      city: [faker.location.city(), faker.location.city(), faker.location.city()],
      date: ['Feb, 5', 'Jan, 1'],
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('takes the least number of zipped items', () => {
    const template = createTemplate('{{greeting:name}} goes to {{greeting:city}}!')
    const data = getTemplateData(template, {
      name: [faker.person.firstName(), faker.person.firstName(), faker.person.firstName()],
      city: [faker.location.city(), faker.location.city()],
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('resolves path properties', () => {
    const template = createTemplate('Hello {{name.first}} {{name.last}}!')
    const data = getTemplateData(template, {
      name: {
        first: faker.person.firstName(),
        last: faker.person.lastName(),
      },
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('resolves array path properties preserving structure of data', () => {
    const template = createTemplate(
      '{{cities.residents.name.first}} {{cities.residents.name.last}} of age {{cities.residents.age}} lives in {{cities.name}}',
    )
    const data = getTemplateData(template, {
      cities: faker.datatype.array(2).map(() => ({
        name: faker.location.city(),
        residents: faker.datatype.array(2).map(() => ({
          name: {
            first: faker.person.firstName(),
            last: faker.person.lastName(),
          },
          age: faker.number.int({ min: 18, max: 65 }),
        })),
      })),
    })
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
  it('preserves the substitution field if root property not found in the data', () => {
    const template = createTemplate('Hello {{name}}!')
    const data = getTemplateData(template, {})
    const content = data.map((d) => formatTemplate(template, d)).join('\n')
    expect(content).toMatchSnapshot()
  })
})
