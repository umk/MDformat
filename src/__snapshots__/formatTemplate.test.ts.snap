// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template creates template with single array data field 1`] = `
"Hello Eduardo!
Hello Leanna!
Hello Benjamin!"
`;

exports[`template creates template with single data field 1`] = `"Hello Eduardo!"`;

exports[`template creates template with three combined array data fields 1`] = `
"Eduardo goes to North Susanview on Feb, 5!
Eduardo goes to Caylafurt on Jan, 1!
Leanna goes to North Susanview on Feb, 5!
Leanna goes to Caylafurt on Jan, 1!"
`;

exports[`template creates template with two permutated array data fields 1`] = `
"Eduardo goes to Fort Haydenmouth!
Leanna goes to Fort Haydenmouth!
Benjamin goes to Fort Haydenmouth!
Eduardo goes to Lake Luzport!
Leanna goes to Lake Luzport!
Benjamin goes to Lake Luzport!"
`;

exports[`template creates template with two zipped array data fields 1`] = `
"Eduardo goes to Fort Haydenmouth!
Leanna goes to Lake Luzport!
Benjamin goes to Felicitystad!"
`;

exports[`template creates template without data 1`] = `"Amor ventosus cur. Asporto rerum sit delego expedita consequuntur. Clam absens pecco laborum concido cohaero adstringo. Acsi ipsa talio certe. Tenetur modi utilis carus ullam."`;

exports[`template doesn't permutate properties, that are not referenced in template 1`] = `
"Hello Eduardo!
Hello Leanna!
Hello Benjamin!"
`;

exports[`template preserves the substitution field if root property not found in the data 1`] = `"Hello {{name}}!"`;

exports[`template resolves array path properties preserving structure of data 1`] = `
"Scotty Stark of age 51 lives in Effietown
Anabel Wuckert of age 62 lives in Effietown
Rusty Sporer of age 53 lives in New Chase
Ansel Murray of age 53 lives in New Chase"
`;

exports[`template resolves path properties 1`] = `"Hello Eduardo Breitenberg!"`;

exports[`template takes the least number of zipped items 1`] = `
"Eduardo goes to Fort Haydenmouth!
Leanna goes to Lake Luzport!"
`;
